File: client-scalaz/src/main/scala/org/openapitools/client/api/Connection.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Connection._

case class Connection (
  id: Option[Long],
created: Option[OffsetDateTime],
updated: Option[OffsetDateTime],
deleted: Option[OffsetDateTime],
objectType: Option[String],
legacyId: Option[String],
searchTags: Option[String],
active: Option[Boolean],
valid: Option[Boolean],
account: Option[Account],
connection: Option[Account],
pendingConnection: Option[ConnectedFriend],
friend: Option[Boolean],
friendRequested: Option[Boolean],
friendRequestPending: Option[Boolean],
following: Option[Boolean],
follower: Option[Boolean],
trusted: Option[Boolean],
blocked: Option[Boolean],
friendRequestDate: Option[OffsetDateTime],
friendResponseDate: Option[OffsetDateTime],
display: Option[String],
connectionViewedDate: Option[OffsetDateTime],
connectionType: Option[ConnectionType],
title: Option[String],
pendingConnectionId: Option[Long],
latitude: Option[Double],
longitude: Option[Double])

object Connection {
  import DateTimeCodecs._
  sealed trait ConnectionType
  case object PRIMARY extends ConnectionType
  case object SECONDARY extends ConnectionType
  case object DEPENDENT extends ConnectionType
  case object ACQUAINTANCE extends ConnectionType

  object ConnectionType {
    def toConnectionType(s: String): Option[ConnectionType] = s match {
      case "PRIMARY" => Some(PRIMARY)
      case "SECONDARY" => Some(SECONDARY)
      case "DEPENDENT" => Some(DEPENDENT)
      case "ACQUAINTANCE" => Some(ACQUAINTANCE)
      case _ => None
    }

    def fromConnectionType(x: ConnectionType): String = x match {
      case PRIMARY => "PRIMARY"
      case SECONDARY => "SECONDARY"
      case DEPENDENT => "DEPENDENT"
      case ACQUAINTANCE => "ACQUAINTANCE"
    }
  }

  implicit val ConnectionTypeEnumEncoder: EncodeJson[ConnectionType] =
    EncodeJson[ConnectionType](is => StringEncodeJson(ConnectionType.fromConnectionType(is)))

  implicit val ConnectionTypeEnumDecoder: DecodeJson[ConnectionType] =
    DecodeJson.optionDecoder[ConnectionType](n => n.string.flatMap(jStr => ConnectionType.toConnectionType(jStr)), "ConnectionType failed to de-serialize")

  implicit val ConnectionCodecJson: CodecJson[Connection] = CodecJson.derive[Connection]
  implicit val ConnectionDecoder: EntityDecoder[Connection] = jsonOf[Connection]
  implicit val ConnectionEncoder: EntityEncoder[Connection] = jsonEncoderOf[Connection]
}

File: client-clojure/src/sirqul_io_t_platform/specs/connection.clj
(ns sirqul-io-t-platform.specs.connection
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.connected-friend :refer :all]
            )
  (:import (java.io File)))


(def connection-data
  {
   (ds/opt :id) int?
   (ds/opt :created) inst?
   (ds/opt :updated) inst?
   (ds/opt :deleted) inst?
   (ds/opt :objectType) string?
   (ds/opt :legacyId) string?
   (ds/opt :searchTags) string?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :account) account-spec
   (ds/opt :connection) account-spec
   (ds/opt :pendingConnection) connected-friend-spec
   (ds/opt :friend) boolean?
   (ds/opt :friendRequested) boolean?
   (ds/opt :friendRequestPending) boolean?
   (ds/opt :following) boolean?
   (ds/opt :follower) boolean?
   (ds/opt :trusted) boolean?
   (ds/opt :blocked) boolean?
   (ds/opt :friendRequestDate) inst?
   (ds/opt :friendResponseDate) inst?
   (ds/opt :display) string?
   (ds/opt :connectionViewedDate) inst?
   (ds/opt :connectionType) string?
   (ds/opt :title) string?
   (ds/opt :pendingConnection_id) int?
   (ds/opt :latitude) float?
   (ds/opt :longitude) float?
   })

(def connection-spec
  (ds/spec
    {:name ::connection
     :spec connection-data}))
